### PR TITLE
fix Alter Table tlbname Alter Column Type with multi indexes reuse index error #16805

### PR DESCRIPTION
**Pull request by HelloYJohn**
Fix https://github.com/greenplum-db/gpdb/issues/16805

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Review by huansong — COMMENTED** [review]
Thanks for the report and fix! It was an oversight in https://github.com/greenplum-db/gpdb/commit/c72f1c5a950b where I missed the case of reusing multiple indexes. 

The fix works. But a slightly better one is to skip early in the `for` loop:
```
@@ -13819,6 +13819,10 @@ ATPostAlterTypeParse(Oid oldId, Oid oldRelId, Oid refRelId, char *cmd,

                        stmt = (IndexStmt *) cmd->def;

+                       /* skip if it's not the same index */
+                       if (stmt->indexOid != oldId)
+                               continue;
+
                        /* if we are not reusing this index, continue */
                        if (!OidIsValid(stmt->oldNode))
                                continue;
```
Performance-wise they're not much different, as we would open the index relation exactly once either way. But to skip early feels better logically.

We can also place an Assert to make sure:
```
@@ -13844,6 +13848,8 @@ ATPostAlterTypeParse(Oid oldId, Oid oldRelId, Oid refRelId, char *cmd,
                                irel = index_open(indoid, AccessShareLock);
                        }

+                       Assert(irel && irel->rd_rel->oid == stmt->indexOid);
+
                        /* replace it with my own */
                        stmt->oldNode = irel->rd_node.relNode;
```

Also, could you please add commit message and PR description? Feel free to use and adjust this draft:
```
This commit fixes an issue introduced by https://github.com/greenplum-db/gpdb/commit/c72f1c5a950b
where we let QE reuse its own index filenode instead of QD's. However, it
did not take multiple indexes into account, so a QE could reuse an
index filenode for another index, which is incorrect. Now fix it by only
reusing filenode of the same index.
```

**Comment by huansong** [comment]
> So is this data incorrectness issue then?

Yes. 

In the reported case, `DROP TABLE` would try to drop the same filenode twice, which might've caused the segment to stuck somewhere so there's prob to have problem during 2PC. But the issue could be manifested in other ways which looks like data incorrectness:

```
postgres=# create table bug(a int);
CREATE TABLE
postgres=# create index bugi on bug(a);
CREATE INDEX
postgres=# create unique index bugui on bug(a);
CREATE INDEX
postgres=# alter table bug alter column a type int;
ALTER TABLE
postgres=# insert into bug values(1);
ERROR:  duplicate key value violates unique constraint "bugui"  (seg1 127.0.1.1:7003 pid=1050)
DETAIL:  Key (a)=(1) already exists.
```
As seen above, it complains about duplicated key w/o real duplication...

**Comment by HelloYJohn** [comment]
> Thanks for the report and fix! It was an oversight in [c72f1c5a950b](https://github.com/greenplum-db/gpdb/commit/c72f1c5a950b) where I missed the case of reusing multiple indexes.
> 
> The fix works. But a slightly better one is to skip early in the `for` loop:
> 
> ```
> @@ -13819,6 +13819,10 @@ ATPostAlterTypeParse(Oid oldId, Oid oldRelId, Oid refRelId, char *cmd,
> 
>                         stmt = (IndexStmt *) cmd->def;
> 
> +                       /* skip if it's not the same index */
> +                       if (stmt->indexOid != oldId)
> +                               continue;
> +
>                         /* if we are not reusing this index, continue */
>                         if (!OidIsValid(stmt->oldNode))
>                                 continue;
> ```
> 
> Performance-wise they're not much different, as we would open the index relation exactly once either way. But to skip early feels better logically.
> 
> We can also place an Assert to make sure:
> 
> ```
> @@ -13844,6 +13848,8 @@ ATPostAlterTypeParse(Oid oldId, Oid oldRelId, Oid refRelId, char *cmd,
>                                 irel = index_open(indoid, AccessShareLock);
>                         }
> 
> +                       Assert(irel && irel->rd_rel->oid == stmt->indexOid);
> +
>                         /* replace it with my own */
>                         stmt->oldNode = irel->rd_node.relNode;
> ```
> 
> Also, could you please add commit message and PR description? Feel free to use and adjust this draft:
> 
> ```
> This commit fixes an issue introduced by https://github.com/greenplum-db/gpdb/commit/c72f1c5a950b
> where we let QE reuse its own index filenode instead of QD's. However, it
> did not take multiple indexes into account, so a QE could reuse an
> index filenode for another index, which is incorrect. Now fix it by only
> reusing filenode of the same index.
> ```

I think it doesn't work, stmt->indexOid hasn't been assigned. The type of stmt is struct IndexStmt(ParseTree), You can debug it
(gdb) f
#0  ATPostAlterTypeParse (oldId=20379, oldRelId=20374, refRelId=0, 
    cmd=0x7e808b8 "CREATE INDEX float2double_table_idx_c1c2c3 ON public.float2double_table USING btree (c1, c2, c3)", wqueue=0x7fffe7e412c0, lockmode=8, rewrite=false)
    at tablecmds.c:13826
13826                           if (irel == NULL)
(gdb) p oldId
$5 = 20379
(gdb) p stmt.indexOid 
$6 = 0
(gdb)

**Comment by huansong** [comment]
> stmt->indexOid hasn't been assigned

Ah yes, my bad. Since we are creating an index here, all we have is the index name. I'm good with the current fix then.

File: src/backend/commands/tablecmds.c [hunk 1]
```diff
@@ -13845,7 +13845,8 @@ ATPostAlterTypeParse(Oid oldId, Oid oldRelId, Oid refRelId, char *cmd,
 			}
 
 			/* replace it with my own */
-			stmt->oldNode = irel->rd_node.relNode;
+			if (strcmp(stmt->idxname, irel->rd_rel->relname.data) == 0)
+				stmt->oldNode = irel->rd_node.relNode;
```
Comment (huansong) [inline src/backend/commands/tablecmds.c:13849]: Please adjust the comment above, e.g.:
```
/* If it is for the current index, replace the relnode with my own. */
```

**Review by huansong — APPROVED** [review]
LGTM! Please do fill the commit message & PR description in order for others to easily understand. Thanks!

Dev pipeline: https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/pr-16806

File: src/test/regress/expected/alter_table.out [hunk 2]
```diff
@@ -4787,3 +4787,11 @@ DROP TABLE table_issue_15494;
 CREATE TABLE IF NOT EXISTS table_issue_15494(c0 boolean);
 ALTER TABLE table_issue_15494 ALTER c0 SET DEFAULT ((1.5::FLOAT) NOTNULL);
 DROP TABLE table_issue_15494;
+-- please refer to:  https://github.com/greenplum-db/gpdb/issues/16805
+create table float2double_table(c1 float,c2 float,c3 float);
```
Comment (huansong) [inline src/test/regress/expected/alter_table.out:4791]: There's a test diff here:
```
--- /tmp/build/e18b2f02/gpdb_src/src/test/regress/expected/alter_table.out	2023-12-05 18:09:38.353932629 +0000
+++ /tmp/build/e18b2f02/gpdb_src/src/test/regress/results/alter_table.out	2023-12-05 18:09:38.577954306 +0000
@@ -4702,7 +4703,7 @@
 ALTER TABLE table_issue_15494 ALTER c0 SET DEFAULT ((1.5::FLOAT) NOTNULL);
 DROP TABLE table_issue_15494;
 -- please refer to:  https://github.com/greenplum-db/gpdb/issues/16805
-create table float2double_table(c1 float,c2 float,c3 float);
+create table IF NOT EXISTS float2double_table(c1 float,c2 float,c3 float);
```

**Comment by HelloYJohn** [comment]
I don't think it's my problem, I see compat-report-75c621bb70e09308529373c340fb8b251a871710.zip.
https://github.com/greenplum-db/gpdb/suites/18792025610/artifacts/1095271734

Could you help me to solve the problem?

**Comment by higuoxing** [comment]
> I don't think it's my problem, I see compat-report-75c621bb70e09308529373c340fb8b251a871710.zip. https://github.com/greenplum-db/gpdb/suites/18792025610/artifacts/1095271734
> 
> Could you help me to solve the problem?

Hi, the problem is fixed in https://github.com/greenplum-db/gpdb/commit/914515d08c88c8268addd40955b1b84bbedf220d

You can fix your branch by rebasing your change against the latest main branch.

**Review by huansong — APPROVED** [review]

**Pull request merged**